src/circuit/gadget/utilities/decompose_running_sum.rs

```diff
@@ -0,0 +1,391 @@
+        self.decompose(region, offset, z_0, strict, word_num_bits, num_windows)
+    }
+
+    /// `z_0` must be the cell at `(self.z, offset)` in `region`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if there are too many windows for the given word size.
+    fn decompose(
+        &self,
+        region: &mut Region<'_, F>,
+        offset: usize,
+        z_0: CellValue<F>,
+        strict: bool,
+        word_num_bits: usize,
+        num_windows: usize,
+    ) -> Result<(CellValue<F>, RunningSum<F>), Error> {
+        // Make sure that we do not have more windows than required for the number
+        // of bits in the word. In other words, every window must contain at least
+        // one bit of the word (no empty windows).
+        //
+        // For example, let:
+        //      - word_num_bits = 64
+        //      - WINDOW_NUM_BITS = 3
+        // In this case, the maximum allowed num_windows is 22:
+        //                    3 * 22 < 64 + 3
+        //
+        assert!(WINDOW_NUM_BITS * num_windows < word_num_bits + WINDOW_NUM_BITS);
+
+        // Enable selectors
+        {
+            for idx in 0..num_windows {
+                self.q_range_check.enable(region, offset + idx)?;
+            }
+
+            if strict {
+                // Constrain the final running sum output to be zero.
+                self.q_strict.enable(region, offset + num_windows)?;
+            }
+        }
+
+        // Decompose base field element into K-bit words.
+        let words: Vec<Option<u8>> = {
+            let words = z_0
+                .value()
+                .map(|word| decompose_word::<F>(word, word_num_bits, WINDOW_NUM_BITS));
+
+            if let Some(words) = words {
+                words.into_iter().map(Some).collect()
+            } else {
+                vec![None; num_windows]
+            }
+        };
+
+        // Initialize empty vector to store running sum values [z_1, ..., z_W].
+        let mut zs: Vec<CellValue<F>> = Vec::with_capacity(num_windows);
+        let mut z = z_0;
+
+        // Assign running sum `z_i`, i = 0..=n, where z_{i+1} = (z_i - k_i) / (2^K).
+        // Outside of this helper, z_0 = alpha must have already been loaded into the
+        // `z` column at `offset`.
+        let offset = offset + 1;
+
+        let two_pow_k_inv = F::from_u64(1 << WINDOW_NUM_BITS as u64).invert().unwrap();
+        for (idx, word) in words.iter().enumerate() {
+            // z_next = (z_cur - word) / (2^K)
+            let z_next = {
+                let word = word.map(|word| F::from_u64(word as u64));
+                let z_next_val = z
+                    .value()
+                    .zip(word)
+                    .map(|(z_cur_val, word)| (z_cur_val - word) * two_pow_k_inv);
+                let cell = region.assign_advice(
+                    || format!("z_{:?}", idx + 1),
+                    self.z,
+                    offset + idx,
```

Comment:
```suggestion
                    || format!("z_{:?}", i + 1),
                    self.z,
                    offset + i,
```
Also consider using `offset + i + 1` and omitting the `let offset = offset + 1;` above.